src/t8_cmesh.hxx

```diff
@@ -31,6 +31,59 @@
 #include <t8_cmesh/t8_cmesh_types.h>
 #include <t8_geometry/t8_geometry_handler.hxx>
 
+/**
+ * Compute the first element of a process in a partitioned mesh, via floor(process * global_num_elements / mpisize).
+ * Prevents overflow by splitting global_num_elements into two 32 bit integers.
+ * global_num_elements = a_0 * tau + a_1, tau = 2^32.
+ * 
+ * floor(a_0 * tau + a_1) * process / mpisize
+ * = floor (a_0 * tau * process / mpisize + a_1 * process / mpisize)
+ * (add the fractional part of the summands and take the floor) to split the computation into separate parts.
+ * = floor (a_0 * tau * process / mpisize) + floor (a_1 * process / mpisize) + floor ((a_o*tau*process) % mpisize + a_1*process % mpisize) / mpisize)
+ * the first term can computed as:
+ * floor (a_0 * tau * process / mpisize)
+ * = floor (floor (a_0 * tau / mpisize * process ) + (a_0 * tau % mpisize) * process / mpisize)
+ * We divide a_0 * tau by mpisize first to prevent an overflow. The second term corrects for the rounding error.
+ *
+ * Adding all summands gives the result.
+ *
+ * This enables us to partition 2^64-1 elements over 2^32-1 processes.
+ * Update this function if we have supercomputers with more than 2^32-1 processes, or need larger meshes.
+ *
+ * \param[in] process   The number of processes
+ * \param[in] mpisize   The size of the MPI communicator
+ * \param[in] global_num_elements   The number of elements in the global mesh
+ * \return The first element of the process in the partitioned mesh
+ * 
+ * \warning This function assumes that process <= mpisize. mpisize has to be greater than 0.
+ * Otherwise the result of process * global_num_elements / mpisize can exceed the range of uint64_t.
+ */
+constexpr uint64_t
+t8_cmesh_get_first_element_of_process (const uint32_t process, const uint32_t mpisize,
+                                       const uint64_t global_num_elements)
+{
+  T8_ASSERT (mpisize > 0);
+  T8_ASSERT (process <= mpisize);
+
+  /* Cast everything into uint64_t */
+  const uint64_t process_64 = static_cast<uint64_t> (process);
+  const uint64_t mpisize_64 = static_cast<uint64_t> (mpisize);
```

Comment:
What is the need for these variables?

Reply:
This is to ensure computation in 64-bit integers. I did not want to cast it everytime so I added these variables. In my opinion this is safer than to rely on promotion and more readable, too.